### ISSUE TITLE
@PolyNull

### ISSUE THREAD

**Issue by kevinb9n** [jspecify/jspecify#79]
Informally, and approximately: a method or constructor with any occurrences of `@PolyNull` in its signature is treated as if it is two distinct overloads: one having all occurrences of `@PolyNull` replaced with `@Nullable`, the other having all replaced with `@NotNull` (under the fantasy that nullness information can actually be used for overload resolution.)

All the cases I'm aware of this being useful involve applying `@PolyNull` to at least one in-type (e.g parameter type) and at least one out-type (e.g. return type).

The alternative (if this annotation is not available) is that these parameters must be marked `@Nullable`; it will be like the previous situation but with the `@NotNull` overload missing. Consumers passing in a not-null expression (which should generally be common) will get a non-nullable result that the compiler will not *know* is non-nullable.

A bit of early discussion was in #3.

---

_cpovirk edit: various cases / design questions to consider if we end up including `@PolyNull`:_

- (probably not [receiver parameters](https://checkerframework.org/manual/#class-qual-param-method))
- usage in invalid locations (e.g., on fields or classes): [typetools/checker-framework#4480](https://github.com/typetools/checker-framework/issues/4480) / [typetools/checker-framework#1376](https://github.com/typetools/checker-framework/issues/1376)
- whether it means ["`UNION_NULL` or `MINUS_NULL`" or rather "`UNION_NULL` or `NO_CHANGE`"](https://github.com/jspecify/jspecify/issues/79#issuecomment-1587719162)
- whether a `@PolyNull` method like `Class.cast` maps `UNSPECIFIED` to `UNSPECIFIED` and, if so, how we'd specify that, including when the annotation is applied to a type variable that would otherwise have unspecified nullness
   - Also, do we need to do anything special [when the type parameter has an unspecified bound](https://github.com/jspecify/jspecify/issues/79#issuecomment-1557692942)?
- whether users would write `@PolyNull @Nullable`, just `@PolyNull`, `@Nullable(onlyJointly = true)`, a [contract](https://github.com/jspecify/jspecify/issues/757), or what (plus related naming questions)
  - A contract would likely limit us to supporting relationships between a parameter's root type and the method's return type's root type, for better (avoids the field/class mixup noted above, simpler to specify and implement) or for worse (cuts off some use cases).
- how it interacts with overrides, including with contravariant parameters, and maybe [subtyping more generally](https://github.com/jspecify/jspecify/issues/79#issuecomment-2470697270)
   - example: `ImmutableMap.getOrDefault` could use `@PolyNull`, but plain `Map.getOrDefault` might or might not be able to, depending on how we resolve the `MINUS_NULL`/`NO_CHANGE` question above

**Comment by cpovirk** [comment]
I was just talking on #32 about using an annotation element on `@Nullable` like `@Nullable(unspecified = true)`. We could consider doing the same for `@PolyNull`.

To be concrete, let's look at `Class.cast`. Here's a version with `@PolyNull`:

```java
class Class<T> {
  @PolyNull T cast(@PolyNull Object o);
}
```

Maybe we'd want to do that with an element on `@Nullable` instead:

```java
class Class<T> {
  @Nullable(onlyJointly = true) T cast(@Nullable(onlyJointly = true) Object o);
}
```

Aside: If a tool doesn't understand `onlyJointly` (perhaps because it wasn't present in jspecify 1.0), then this approach degrades somewhat gracefully. That's kind of nice.

Another nice feature is that this approach somewhat "hides" `@PolyNull`, just as I've talked about ["hiding" `@NullnessUnspecified`](https://github.com/jspecify/jspecify/issues/32#issuecomment-613020539).

I'm about to post more about the idea of annotation elements in general on #32.

**Comment by kevinb9n** [comment]
Solving this might be a little more important than I was first thinking, because it should mean that `Map.getOrDefault` can become an actually useful remedy for the problems with `Map.get`.

**Comment by kevin1e100** [comment]
(I will admit rarely using getOrDefault to solve such problems.  In my own
experience the primary "solution" is to iterate a map's entries() instead
of keys(), thereby avoiding map.get() in the loop.)

**Comment by cpovirk** [comment]
I have a similar reaction to @kevin1e100 . Even in the non-iteration case, I'm not sure how often `getOrDefault` is helpful.

Yes, sometimes I can call `map.get` and treat an absent key just like it was present with a default value. For example, if my `Map<String, List<Object>>` is missing a key, that's probably much like mapping it to an empty list.

But often what I want is to do something different entirely: Maybe it's to throw an exception, to look up the value elsewhere, etc.

In that case, I *can* switch to `getOrDefault` with a dummy value. But I still need an `if` statement, so I'm not much better off. Arguably I'm actually *worse* off, since my null-checking tool won't *force* me to handle the "not present" case anymore, so I might accidentally pass the dummy value through many layers of my program before it's detected.

(Not to mention that I can't even employ this approach in the first place unless there's some distinguished value I can pass as the default, knowing that it will never appear as a true value.)

Now, all this is just my memory of anecdotal experience. If we want better data, we could do some searches over our codebase.

**Comment by cpovirk** [comment]
Additionally: To support the cases in which `getOrDefault` *is* helpful, we can consider the other sound way of annotating the method:

`V getOrDefault(@Nullable Object key, V defaultValue)`

(This is roughly [what Checker Framework does] (https://github.com/typetools/jdk/blob/bd70449a7b5049b57ef3467c4c7a3f3de108ca1e/src/java.base/share/classes/java/util/Map.java#L648).)

Of course, the signature forbids some valid calls. So this leads us back to larger questions about the "right" way to annotate APIs. My current view is that, just as there's not always one definitively "right" way to write a plain Java API today, there's not always one definitively "right" way to add nullness annotations, either. It can be a trade-off, and I think both possible `getOrDefault` signatures deserve some consideration.

**Comment by cpovirk** [comment]
> there's not always one definitively "right" way to write a plain Java API today

We can actually take methods like `Collection.contains` as an example of this: The API is `contains(Object)`, and there are [reasonable](https://smallwig.blogspot.com/2007/12/why-does-setcontains-take-object-not-e.html) [arguments](https://errorprone.info/bugpattern/CollectionIncompatibleType#footnote-would-requiring-e-have-been-better) for that, but other languages get by with `contains(E)`, which prevents a class of bugs.

That said, when we make such decisions for nullness in particular, we always want to keep in mind that Kotlin might outright _prevent_ passing a nullable value for a non-nullable parameter. But that is probably tolerable for `getOrDefault(..., nullableDefaultValue)`, since users can still call `get` as needed. (And creative users can do something like `Collections.<@Nullable Foo, Bar>unmodifiableMap(map).get(nullableFoo)` :))

**Comment by kevin1e100** [comment]
Interesting point that being able to do getOrDefault(key, null) isn't
super-valuable, since that's the same as get(key) only more expensive.

**Comment by kevinb9n** [comment]
Note that as we struggle to make Google's massive internal codebase more refactorable and evolvable, we are tackling more and more "type migrations" where one type (say, Date) must be replaced with another (say, Instant). In doing this we have to spend a lot of time in an intermediate state where conversion methods (forward or backward) are being called all over the place.

It is just *terrifically* convenient to keep these conversion methods orthogonal to nullness; i.e., always convert null to null and non-null to non-null. (When we can't That means they are `@PolyNull` methods. If we don't have `@PolyNull` we will have to habitually make TWO differently-named conversion methods in each direction for any type migration.

At the moment I don't see how long we can be happy without `@PolyNull`.

**Comment by kevinb9n** [comment]
This is just one opinion (from the internets), but users calling something their "biggest pain point" (after converting 400 KLOC!) is significant even in small numbers.

"The biggest pain point with Eclipse NP analysis is its lack of PolyNull."

https://www.reddit.com/r/java/comments/obscjo/handling_null_and_upgrading_past_java_8_inside/h3pso50/?utm_source=reddit&utm_medium=web2x&context=3

**Comment by kevinb9n** [comment]
Collecting up thoughts from the above.

* Indeed in our current model the meaning of `@PolyNull` should be "either all are union-null, or all are just as-is".

* It seems like `@PolyNull` can only make sense when used on at least one in-type and at least one out-type.

* The name `@JointlyNullable` seems to convey more information than `@PolyNull`, but will any user intuit what this means without reading a sentence worth of javadoc anyway? To me the compact name `@PolyNull` seems to work.

* I feel doubtful that boolean parameters like `@Nullable(jointly = true)` will work out nicely. All it can do is add more semantics onto what the annotation itself already captures, but that just seems more cleanly expressed as two annotations, or as one annotation that `@Implies` the other. We could dissect this more, but intuitively I feel that a forest of separate annotations with various implies-links will be flexible and evolvable, whereas one annotation with a few parameters seems to lock a bunch of stuff together in an unnecessarily rigid way. Again, intuition not assertion.

* If we don't do this, then the things we would have made `@PolyNull` will have to be `@Nullable`, yes?

* And if we do it, but Tool X doesn't support it correctly, how would we expect it to degrade? Given the previous, it seems like it should treat those types as `@Nullable`. You won't have a problem calling the method, but you might have lost some information about what you get back.

**Comment by cpovirk** [comment]
I agree that a Java nullness checker needs to contain `@PolyNull` or similar functionality (at least for a small number of hardcoded methods) to be practically usable at large scale. While I feel OK having checkers hardcode methods in the short term, I expect that that will mostly just give us time to:

- confirm just how many distinct methods need this handling (and the type-migration use case has further increased what I expect us to find)
- catch up on formalizing, firming up, and documenting our semantics thus far and continue on to doing the same for `@PolyNull` (possibly including [aliasing/implies](https://github.com/jspecify/jspecify/issues/35), which I mention as a possible prerequisite at the end of this post)

As for naming: My only fear with the name "PolyNull" is that people may reasonably assume that it has the behavior of the existing Checker Framework `@PolyNull`. (The naming precedent runs even deeper than that, extending beyond nullness and into the Checker Framework's other type systems, as in [`@PolyPresent`](https://checkerframework.org/manual/#optional-annotations).) It would be particularly sad for Checker Framework users to see subtly different behavior depending on which `@PolyNull` they had imported.

(As for `@Nullable(jointly = true)`, I am hoping that I brought that up largely for completeness after a comment about boolean parameters in #32 :) I'm at least not a fan now -- and possibly not a fan of annotation parameters [in general](https://github.com/jspecify/jspecify/issues/109#issuecomment-630517056).)

And for the no-`@PolyNull` world, my thinking has matched your thoughts above:

- People should typically use `@Nullable` if `@PolyNull` isn't available to them.
- `@PolyNull` should behave like `@Nullable` for tools that don't understand it. (That may mean that we want to complete support for `@Implies(Nullable.class)` before completing this. But if we really wanted, we could get by without: We could release a version of JSpecify _without_ the `PolyNull.class` but with a spec that covers it, and then we could release a _later_ version that actually adds the annotation after tools are ready to handle it -- or at least ready to treat it like `@Nullable`.)

("Treat it like `@Nullable`" may lead to some unfortunate behavior in edge cases: If you _implement_ a method with `@PolyNull` in its signature, then a "treat it like `@Nullable`" tool will let you return `null` even if the input wasn't `null`. But it seems clear that that's preferable to assuming that the method _never_ accepts or returns `null`. And those are our only real options, barring abuse of unspecified nullness.)

**Comment by kevinb9n** [comment]
(to the last bit, a reminder we have explained in our requirements doc that the needs of API consumers have precedence over those of implementors in case of conflict)

**Comment by cpovirk** [comment]
True. There are probably ways for _consumers_ to suffer, too, like having to suppress when calling this contrived method:

```java
void rebalance(List<@PolyNull String> left, List<@PolyNull String> right);
```

But again, that's contrived. And it's a false positive rather than a false negative. (Maybe there's a way to produce a false negative, but I haven't thought about it much.)

Anyway, this may be another case in which it's hard to see a true conflict between consumers and implementors. I think our principle of "If you want nulls to pass through, then it needs to be `@Nullable`" continues to serve both use cases well.

**Comment by cpovirk** [comment]
Trying to categorize use cases into "more compelling" and "less compelling"

More compelling:
- conversions: `@PolyNull Foo convert(@PolyNull Bar bar)`
- default values: `@PolyNull Foo getOrDefault(K key, @PolyNull Foo defaultValue)`

Less compelling:
- `List<@PolyNull Foo>` can typically be `List<? extends @Nullable Foo>` (which also offers more flexibility to people who don't even use nullness checking).
- `<T> @PolyNull T passthrough(@PolyNull T value)` can be `<T extends @Nullable Object> T passthrough(T value)`.
- A method that uses `@PolyNull` only once typically doesn't need it (though there are occasional exceptions)

My very rough sense from a quick skim of usages in Google's codebase is that ~2/3 fall into the "more compelling" bucket (mostly "conversions") and ~1/3 fall into the "less compelling" bucket (mostly `List<@PolyNull Foo>`). I would need to look more to say with confidence, but that should give a general sense.

**Comment by cpovirk** [comment]
I have a few other things I've been considering saying here, but I'll limit myself to 2 :)

1. [KT-8889](https://youtrack.jetbrains.com/issue/KT-8889) is the Kotlin feature request for "if a given function parameter is not null, the result is not null." It links to issues about `@Contract`, which I have suggested is [not as flexible as `@PolyNull` could be](https://github.com/jspecify/jspecify/issues/79#issuecomment-573820048).

2. One interesting feature of JSpecify is "unspecified nullness." Above, we've been talking about treating `@PolyNull` as "`@Nullable` or not." We would eventually want to think more about whether the "right" treatmeant is "`@Nullable`, unspecified nullness, or neither." If we had only "`@Nullable` or not," then lenient tools would probably consider a call to silently convert an unspecified-nullness input to a _non-null_ output, and strict tools would probably consider a call to silently convert an unspecified-nullness input to a _nullable_ output. Maybe that's tolerable?

**Comment by kevinb9n** [comment]
Your answers to my last couple July 2 questions seem encouraging to me:  this seems to suggest that there can be a reasonably well-behaved JSpecify *without* a `@PolyNull` annotation, and either a future version of JSpecify or just a checker-specific artifact could provide a `@PolyNull` that has `@Implies(Nullable.class)`. Then APIs that want to adopt that can replace some usages of `@Nullable` with it, and tools that don't recognize it will simply follow the "implies" link.

If this is correct, it's a win, I think.

**Comment by cpovirk** [comment]
#214 raises the idea of `@Nullable("s")`. I have three thoughts:

First: That could be used to support "unrelated" pairs of `@PolyNull` types in the same signature. For example, here's a real-world edge case from Google's codebase:

```
class BiOptional<A, B> {
  Pair<@PolyNull A, @PolyNull B> orElse(@PolyNull A a, @PolyNull B b);
}
```

In that example, it would be nice to be able to replace the 4 usages of `@PolyNull` with 2 usages of `@Nullable("a")` and 2 usages of `@Nullable("b")`.

Second: We should think about how this interacts with [aliasing/implies](https://github.com/jspecify/jspecify/issues/35). As we've noted in [other discussions](https://github.com/jspecify/jspecify/issues/35#issuecomment-625969158), we're not sure if there will be a good way to say that one annotation implies another when annotation elements (like `"a"` here) are involved.

Third: I worry that `@Nullable("a")` alone won't make clear to users what the `"a"` means. Maybe this is solvable by picking a name for the attribute other than the default name "`value`," thereby forcing users to write `@Nullable(condition = "a")` or similar -- though of course that is longer.

**Comment by kevinb9n** [comment]
> #214 raises the idea of `@Nullable("s")`.

With two seconds' thought, it sounds like we have a pretty smooth ramp toward that, i.e. if we released a `@PolyNull` without that we could still add it later. If that sounds right, then to the extent we want to discuss that parameter idea, let's make it a separate issue.

**Comment by cpovirk** [comment]
Here's an example where we'd put `@PolyNull` on zillions of APIs -- far more than I'd seen [during my first look](https://github.com/jspecify/jspecify/issues/79#issuecomment-548889923):

For `map` fields, the protobuf compiler [generates](https://developers.google.com/protocol-buffers/docs/reference/java-generated#map-fields) a method like this:

```
Foo getFooOrDefault(Key key, Foo defaultValue);
```

And default values [are one of the canonical use cases](https://github.com/jspecify/jspecify/issues/79#issuecomment-897692415), so I'd want it to generate:

```
@PolyNull Foo getFooOrDefault(Key key, @PolyNull Foo defaultValue);
```

Users do in fact pass `null` as a default fairly commonly, so I wouldn't want to leave the type as non-null. But they more frequently pass non-null values, so a `@Nullable` return type will get annoying after a while.

**Comment by cpovirk** [comment]
protobuf followup: I am a little surprised to see very few errors in our codebase if I just use `@Nullable` instead of `@PolyNull`. That said, I had found that we would see a lot more if we adopted nullness checking more widely.

So I would still expect the _average_ user to get by tolerably well with a checker that just implements special cases for APIs like `Class.cast`. But I also still expect for enough users to need this annotation for us to need to provide it eventually.

**Comment by ben-manes** [comment]
A usage in Caffeine is,
```java
@PolyNull V get(K key, Function<? super K, ? extends @PolyNull V> mappingFunction);
```

This corresponds to `Map.computeIfAbsent`, which internally it delegates to. If the value is not computable, such as a login email being absent in the database (e.g. a typo) then the result is null.

This differs from Guava's cache which instead throws an exception. This was a difficult decision on being null hostile vs real data being absent. While I prefer the null approach, that was settled by the addition of compute methods. Given the Guava's philosophy of being a natural extension to the Collections Framework, Caffeine shifted to that newer style.

A quirk is `LoadingCache.get(key)` which can return null since the `CacheLoader.load(key)` may. There is no way to encode that in an understandable fashion, so initially it was marked as `@Nullable` for correctness and documentation. As tooling has improved that annoyed Kotlin users, so we dropped it to be effectively `@NonNull` per the majority of usages.

`AsyncCache` doesn't have these problems because it computes a `CompletableFuture<V>`. That future being null is a programming bug, but it could resolve to a null value. If so, then the future is considered to have failed (though not exceptionally) and failed future are removed from the cache.

It wouldn't be awful if `Cache.get(key, mappingFunction)` similarly feigned a non-null return value, even if not correct. The vast majority of usages don't take advantage of this property and would instead surface an application exception, e.g. a JAX-RS `BadRequestException` in the failed login example. That wouldn't be the case for Map computations like `compute` and `computeIfPresent` where a null return is often the logics intent (e.g. to synchronously notify a listener on removal).

**Comment by cpovirk** [comment]
I just came across another pattern of `@PolyNull` usages that turns out not to be necessary. I edited it into [my list above](https://github.com/jspecify/jspecify/issues/79#issuecomment-897692415), but to repeat it here:

I saw some methods along the lines of `foo(@PolyNull String string, @PolyNull Listener listener)`. The intent was presumably to require that both values be present or both values be absent. But in fact the methods permit any combination of those, since, e.g., a non-null string and null listener fulfills `@Nullable String` and `@Nullable Listener`. These declarations might as well just be `foo(@Nullable String string, @Nullable Listener listener)`.

**Comment by kevinb9n** [comment]
I believe `@PolyNull` has to appear at least once on an in-type and once on an out-type to make sense.

**Comment by cpovirk** [comment]
> 2\. Above, we've been talking about treating `@PolyNull` as "`@Nullable` or not." We would eventually want to think more about whether the "right" treatment is "`@Nullable`, unspecified nullness, or neither."

For some reason, I was thinking of this again recently. I'll dump my dense notes here for possible future cleanup:

I was tying this question to https://github.com/jspecify/jspecify/issues/69: If I have a type-variable usage `T NO_CHANGE` with the type parameter `T` declared with a bound of `Foo UNSPECIFIED`, then how do I know that `class.cast(t)` should return `UNSPECIFIED`? I can't just ask "Is `T NO_CHANGE` a subtype of `Object UNSPECIFIED`?" because:
- If I require that subtyping to hold in only _some_ world, then _everything_ is a subtype of `Object UNSPECIFIED`, even types like `T UNION_NULL` that we don't want to cover here.
- If I require that subtyping to hold in _all_ worlds, then the check will fail for `T NO_CHANGE`.

How do we "let in" `T NO_CHANGE` but not `T UNION_NULL`? We may need an additional concept of "`T NO_CHANGE` as `Object ___`," which is a little like substitution (but doesn't involve type arguments) and a little like subtyping (but passes through `UNSPECIFIED`).

(This may be important for _getting the types right at all_. That may make it more important than the rest of #69, which is about avoiding some "But why?" warnings or errors.)

**Comment by cpovirk** [comment]
> We may need an additional concept of "`T NO_CHANGE` as `Object ___`," which is a little like substitution (but doesn't involve type arguments) and a little like subtyping (but passes through `UNSPECIFIED`).

Actually, an internal discussion with kevin1e100 just touched on the idea that we might need that for casts, too: If you have an expression of type `T ___` and you cast it to `Foo` (or vice versa?), we might also need to know that? I'm not sure, and I don't have time to think about it today :)

**Comment by cpovirk** [comment]
One thing that we've discussed but seemingly not written about yet here is how Kotlin reduces the need for `@PolyNull`, at least if you design APIs with Kotlin in mind.

### `?.let`

In Kotlin, developers are less likely to define `@PolyNull`-style APIs and more likely to define APIs that accept and return non-nullable types. Then, callers who know that they have a non-nullable type can call the method directly (`foo.bar()`), and callers who know that they have a nullable type can call the method with `?.let` (`foo?.let { it.bar() }`). The compiler ensures that you don't call the method directly on a nullable type (but it does let you call it on a platform type, which can be unfortunate!), yet it ensures that the return type of the `?.let` approach is a nullable type.

### extension functions

In at least the specific case of `Optional.orElse`, Kotlin provides (thanks to [some people](https://github.com/Kotlin/KEEP/blob/master/proposals/stdlib/optional-extensions.md) :)) an extension function ([`getOrDefault`](https://github.com/JetBrains/kotlin/blob/22a3000789e3ea82096f659caaa4326315ef0623/libraries/stdlib/jdk8/src/kotlin/jvm/optionals/Optionals.kt#L22)) that provides `@PolyNull`-like behavior. I think the reason that this works for `Optional.orElse` but doesn't work for `Class.cast` or `Converter.convert` is that the input type and output type are the same type `T`. (It might also be important that `Optional` has its type parameter [declared as non-nullable under our approach](https://github.com/jspecify/jspecify/issues/78). That's the case for `Class` and `Converter`, too, so it's not part of the problem for those classes' methods; it just might be part of the problem for others.)

### other stuff?

I don't know, probably? You could reasonably claim that the various nullness-handling operators (like `?:`) would make it somewhat practical for a Kotlin-focused `Optional` type to choose not to provide `orElse`/`getOrDefault` functionality at all: If it instead provided `orNull`, users could combine that with `?:` to get the other methods' behavior.

**Comment by agentgt** [comment]
BTW speaking of Caffeine

> A quirk is LoadingCache.get(key) which can return null since the CacheLoader.load(key) may. There is no way to encode that in an understandable fashion, so initially it was marked as @Nullable for correctness and documentation. As tooling has improved that annoyed Kotlin users, so we dropped it to be effectively @NonNull per the majority of usages.

It is funny you mention that because this confused me when going to see if we were using Caffeines getOrDefault.

```java
	private static final LoadingCache<CacheKey, ParsedSql> cache = Caffeine.newBuilder()
		.maximumSize(100)
		.<CacheKey, ParsedSql>build(new CacheLoader<CacheKey, ParsedSql>() {
			@Override
			public @NonNull ParsedSql load(
					CacheKey key)
					throws Exception {
				return _parseSql(key.sql, key.config);
			}

		});

	private static ParsedSql _parseSqlCache(
			String sql,
			SqlPlaceholderParserConfig config) {
		return requireNonNull(cache.get(new CacheKey(config, sql)));
	}
```

If I don't have that `requireNonNull` Eclipse will complain with:

```
Unsafe interpretation of method return type as '@NonNull' based on the receiver type '@NonNull LoadingCache<@NonNull CacheKey,@NonNull ParsedSql>'. Type 'LoadingCache<K,V>' doesn't seem to be designed with null type annotations in mind
```

And it does not because Eclipse still cannot read Checkers default annotation placement:

```java
@DefaultQualifier(value = NonNull.class, locations = TypeUseLocation.FIELD)
@DefaultQualifier(value = NonNull.class, locations = TypeUseLocation.PARAMETER)
@DefaultQualifier(value = NonNull.class, locations = TypeUseLocation.RETURN)
package com.github.benmanes.caffeine.cache;
```

But originally I did not read that error or it indeed changed on version upgrade as Ben alluded to from `@Nullable` to Eclipses unspecified. I just assumed cache.get was always nullable.

The reality is I just haven't gone and externally annotated (EEA) Caffeines type parameters yet.

Anyway I thought that might be just another example of how much the pain of going around annotating existing libraries is and how potentially error prone it can be.

**Comment by ben-manes** [comment]
@agentgt fyi https://github.com/ben-manes/caffeine/issues/594 in [v3.0.5](https://github.com/ben-manes/caffeine/releases/tag/v3.0.5). It seemed reasonable to follow Guava's lead which adopted CheckerFramework's annotations as a replacement to jsr305's, where I had no strong opinion and prefered having those who want to discuss it to come to the like minded Guava/JSpecify folks than to me. 🙂

**Comment by agentgt** [comment]
@ben-manes Yes I agree. I didn't mean to make it out as bad choice because I think the first time around when I did it I was annoyed probably similar to the kotlin devs about `LoadingCache.get` being nullable.

It was a safe choice and did not break anything on our end. My jaded tone was more for the libraries that either do not provide any sort of annotations or incorrectly annotated... (well that and Eclipse but Eclipse kind did ok here).

**Comment by cpovirk** [comment]
Sorry, I'm ignoring the posts above for the moment. I haven't caught up yet, but I wanted to dump an unrelated thought.

I'd forgotten another way that Kotlin can get `@PolyNull`-like behavior: Kotlin lets users declare overloads of a function that differ only in nullness. So you can have `fun cast(o: Object): T` and `fun cast(o: Object?): T?`, at least when you're writing Kotlin.

I should look more at what happens if you write a _Java_ API with one set of nullness annotations in a superclass (and/or interface?) and then a Java override with the other set.... I had _thought_ that Kotlin responded by [interpreting the API as having platform types](https://github.com/jspecify/jspecify/issues/49#issue-471586073), but I thought I might have seen otherwise in some cases. I'd need to look. (Might it matter whether the _return type_ matches?) And even if it does "work," I would be hesitant to rely on what might be a bug, given that the platform-type behavior seemed to be the intended one.

**Comment by cpovirk** [comment]
I also want to dump a link to a Guava discussion in which `@PolyNull` could have been part of the answer but in which I concluded that `@NonNull` would suffice (though tools' type inference might be confused either way): https://github.com/google/guava/issues/6824#issuecomment-1799740219

**Comment by cpovirk** [comment]
Apparently C# took the approach that you can apply their equivalent of `@PolyNull` only to a pair of types that are (a) the return type and (b) one parameter type: https://devblogs.microsoft.com/dotnet/try-out-nullable-reference-types/#nullness-dependence-between-inputs-and-outputs-notnullifnotnullstring

That's a nice simplification if you can get away with it. (I do wonder if they could have let you annotate the _parameter_, instead, so that you wouldn't have to refer to the parameter name with a string?) It definitely does solve the common case of null passthrough, but I was surprised at first glance at how many of the `@PolyNull` usages I looked at were more involved than that. (Compare [my comment about supporting only root types with `@Contract`](https://github.com/jspecify/jspecify/issues/79#issuecomment-573820048).) It remains an option, one that certainly covers lots of _calls_ even if it covers a smaller number of _declarations_. (For example, covering `Class.cast` would get us a lot of mileage.)

(Also, more generally: C# had to do a _lot_ of stuff to support its `ref` and perhaps `out` parameters, plus more for its representation of nullable primitives with `Nullable<int>`, etc. It is convenient for us that Java has avoided those particular things.)

**Comment by kevinb9n** [comment]
For starters I think PolyNull only makes sense if you mark at least one in-type and at least one out-type (see our wonderful glossary doc). So the questions are

* what should happen if they don't? (producing an "inference" there seems logical)
* should we limit either group to just one?

**Comment by cpovirk** [comment]
I recently noted in https://github.com/jspecify/jspecify/issues/705 that `@PolyNull` would need to be covered by rules for type inference. \[edit: See https://github.com/eisop/checker-framework/issues/2030 for an example of complexity there.\]

Today, I'm seeing an example of how it would need to be covered by rules for subtyping... or maybe some higher-level rule? I'm not sure it's as simple as "`@PolyNull` is a subtype of _x_ and a supertype of _y_." It seems more like we'd actually need to check [the two conceptual versions of an annotated method](https://checkerframework.org/api/org/checkerframework/checker/nullness/qual/PolyNull.html) separately, both when checking whether a method is a valid override and when checking whether its implementation is valid (including checking any calls to the supermethod, I guess, and I'm not clear on whether we'd treat the supermethod "as `@PolyNull`" or "as the `@Nullable`/`@NonNull` version that we're currently looking at"). My assumption is that the Checker Framework does things along those lines, since it has a concept of "qualifier polymorphism" that seems to go beyond subtyping. I'm sure we could learn from it as needed.

**Comment by cpovirk** [comment]
One subtlety that's come up in some recent contracts discussions is [another](https://github.com/jspecify/jspecify/issues/79#issuecomment-573820048) small difference between `@PolyNull` and `@Contract("null -> null, !null -> !null)`: That particular `@Contract` in theory lets you refine the parameter type based on the check of the return type, but `@PolyNull` does not. However, as best we can tell, no one implements or particularly wants that kind of refinement.

The basic idea is something like:

```java
String maybeNull = ...;
Integer lengthOrNull = lengthOrNull(maybeNull);
if (lengthOrNull != null) {
  return maybeNull.substring(...);
}
```

That's safe under the stated contract because `lengthOrNull` is null if and only if `maybeNull` is.

`@PolyNull`, by contrast, does not make that guarantee. Here's an example of why not:

```java
@PolyNull T withBackup(@Nullable T primary, @PolyNull T backup)
```

There, `withBackup` could return non-null even when `backup` is null, since it might be returned a non-null value of `primary`.

Not, incidentally, that you should write that signature when when you could instead write:

```java
<T extends @Nullable Object> T withBackup(@Nullable T primary, T backup)
```

But the main point here is that users of `@PolyNull`-like contracts end up leveraging only the `!null -> !null` part of the contract, not the `null -> null` part.
